Library/Homebrew/cmd/cleanup.rb

```diff
@@ -83,7 +83,9 @@ def cleanup_cache
         next
       end
 
-      if f.version > version || ARGV.switch?('s') && !f.installed? || bottle_file_outdated?(f, file)
+      file_is_stale = PkgVersion === version ? f.pkg_version > version : f.version > version
+
+      if file_is_stale || ARGV.switch?('s') && !f.installed? || bottle_file_outdated?(f, file)
```

Comment:
I realise the existing logic is also hairy but might be worth splitting this out into e.g. a few `return`s so it's easier to follow.